neo/Ledger/BlockChain.State.cs

```diff
@@ -0,0 +1,181 @@
+        }
+
+        public StateRootState GetStateRoot(uint index)
+        {
+            return currentSnapshot.StateRoots.TryGet(index);
+        }
+
+        public bool GetStateProof(UInt256 root, StorageKey skey, out HashSet<byte[]> proof)
+        {
+            var trieReadOnlyDb = new TrieReadOnlyStore(Store, Prefixes.DATA_MPT);
+            var readOnlyTrie = new MPTReadOnlyTrie(root, trieReadOnlyDb);
+            return readOnlyTrie.GetProof(skey.ToArray(), out proof);
+        }
+
+        public bool VerifyProof(UInt256 root, byte[] key, HashSet<byte[]> proof, out byte[] value)
+        {
+            var result = MPTTrie.VerifyProof(root, key, proof, out value);
+            if (result) value = value.AsSerializable<StorageItem>().Value;
+            return result;
+        }
+
+        private RelayResultReason OnNewStateRoot(StateRoot state_root)
+        {
+            if (state_root.Index < StateRootEnableIndex || state_root.Index <= StateHeight) return RelayResultReason.Invalid;
+            if (state_root.Witness is null) return RelayResultReason.Invalid;
+            if (stateRootCache.ContainsKey(state_root.Index)) return RelayResultReason.AlreadyExists;
+            if (state_root.Index > Height || (state_root.Index > StateHeight + 1 && state_root.Index != StateRootEnableIndex))
+            {
+                stateRootCache.Add(state_root.Index, state_root);
+                return RelayResultReason.Succeed;
+            }
+            var state_root_to_verify = state_root;
+            var state_roots_to_verify = new List<StateRoot>();
+            while (true)
+            {
+                state_roots_to_verify.Add(state_root_to_verify);
+                var index = state_root_to_verify.Index + 1;
+                if (index > Height) break;
+                if (!stateRootCache.TryGetValue(index, out state_root_to_verify)) break;
+            }
+            foreach (var state_root_verifying in state_roots_to_verify)
+            {
+                using (Snapshot snapshot = GetSnapshot())
+                {
+                    stateRootCache.Remove(state_root_verifying.Index);
+                    if (!state_root_verifying.Verify(snapshot))
+                    {
+                        break;
+                    }
+                    var local_state = snapshot.StateRoots.GetAndChange(state_root_verifying.Index);
+                    if (local_state.StateRoot.Root == state_root_verifying.Root && local_state.StateRoot.PreHash == state_root_verifying.PreHash)
+                    {
+                        snapshot.StateRootHashIndex.GetAndChange().Index = state_root_verifying.Index;
+                        snapshot.StateRootHashIndex.GetAndChange().Hash = state_root_verifying.Hash;
+                        local_state.StateRoot = state_root_verifying;
+                        local_state.Flag = StateRootVerifyFlag.Verified;
+                        if (state_root_verifying.Index + 3 > HeaderHeight)
```

Comment:
Why +3?

Reply:
I think we can remove it, and use LocalNode.RelayDirectly. But I'm not sure where this is correct, I suggest we keep it until @KickSeason back.

Reply:
Only relay the latest 2 stateroot.

Reply:
@KickSeason  I have replaced `SendDirectly` by `RelayDirectly`.